pkg/chunk/aws_storage_client.go

```diff
@@ -382,22 +621,67 @@ func (b dynamoDBWriteBatch) Add(tableName, hashValue string, rangeValue []byte,
 	})
 }
 
-type dynamoDBReadBatch []map[string]*dynamodb.AttributeValue
-
-func (b dynamoDBReadBatch) Len() int {
-	return len(b)
+// Fill 'to' with WriteRequests from 'from' until 'to' has at most max requests. Remove those requests from 'from'.
+func (b dynamoDBWriteBatch) takeReqs(from dynamoDBWriteBatch, max int) {
+	outLen, inLen := b.Len(), from.Len()
+	toFill := inLen
+	if max > 0 {
+		toFill = util.Min(inLen, max-outLen)
+	}
+	for toFill > 0 {
+		for tableName, fromReqs := range from {
+			taken := util.Min(len(fromReqs), toFill)
+			if taken > 0 {
+				b[tableName] = append(b[tableName], fromReqs[:taken]...)
+				from[tableName] = fromReqs[taken:]
+				toFill -= taken
+			}
+		}
+	}
 }
 
-func (b dynamoDBReadBatch) RangeValue(i int) []byte {
-	return b[i][rangeKey].B
+type dynamoDBReadRequest map[string]*dynamodb.KeysAndAttributes
+
+func (b dynamoDBReadRequest) Len() int {
+	result := 0
+	for _, reqs := range b {
+		result += len(reqs.Keys)
+	}
+	return result
 }
 
-func (b dynamoDBReadBatch) Value(i int) []byte {
-	chunkValue, ok := b[i][valueKey]
+func (b dynamoDBReadRequest) Add(tableName, hashValue string, rangeValue []byte) {
+	requests, ok := b[tableName]
 	if !ok {
-		return nil
+		requests = &dynamodb.KeysAndAttributes{
+			AttributesToGet: []*string{aws.String(valueKey)},
+			ConsistentRead:  aws.Bool(true),
+		}
+		b[tableName] = requests
+	}
+	requests.Keys = append(requests.Keys, map[string]*dynamodb.AttributeValue{
+		hashKey:  {S: aws.String(hashValue)},
+		rangeKey: {B: rangeValue},
+	})
+}
+
+// Fill 'to' with WriteRequests from 'from' until 'to' has at most max requests. Remove those requests from 'from'.
+func (b dynamoDBReadRequest) takeReqs(from dynamoDBReadRequest, max int) {
+	outLen, inLen := b.Len(), from.Len()
+	toFill := inLen
+	if max > 0 {
+		toFill = util.Min(inLen, max-outLen)
+	}
+	for toFill > 0 {
+		for tableName, fromReqs := range from {
+			taken := util.Min(len(fromReqs.Keys), toFill)
+			if taken > 0 {
+				b[tableName].Keys = append(b[tableName].Keys, fromReqs.Keys[:taken]...)
+				from[tableName].Keys = fromReqs.Keys[taken:]
+				toFill -= taken
+			}
+		}
```

Comment:
I think you'll have fewer bugs if you make this (and equivalent write version) immutable. i.e. have it return a new request populated from `from` with `max` or more, without changing `b`.